src/hotspot/share/opto/addnode.cpp

```diff
@@ -1004,16 +991,29 @@ const Type *XorINode::add_ring( const Type *t0, const Type *t1 ) const {
   const TypeInt *r0 = t0->is_int(); // Handy access
   const TypeInt *r1 = t1->is_int();
 
-  // Complementing a boolean?
-  if( r0 == TypeInt::BOOL && ( r1 == TypeInt::ONE
-                               || r1 == TypeInt::BOOL))
-    return TypeInt::BOOL;
+  if (r0->is_con() && r1->is_con()) {
+    // just XOR them bits.
+    return TypeInt::make( r0->get_con() ^ r1->get_con() );
+  }
+
+  // not constants
```

Comment:
Well I suppose one of them could still be a constant... You should make your comments here a little more precise ;)